cache/src/main/kotlin/com/dropbox/android/external/cache4/CacheEntry.kt

```diff
@@ -0,0 +1,12 @@
+package com.dropbox.android.external.cache4
+
+/**
+ * A cache entry holds the [key] and [value] pair,
+ * along with the metadata needed to perform cache expiration and eviction.
+ */
+internal data class CacheEntry<Key, Value>(
+    val key: Key,
+    @Volatile var value: Value,
```

Comment:
as far as i can read, these are always accessed inside a lock which will put a memory barrier hence you should not need volatile for these.

Reply:
Hmm I’m not sure I follow. These entries live in the ConcurrentHashMap and a couple of synchronized LinkedHashSet but we are updating these values directly (eg when replacing the value for an existing key) instead of recreating and putting a new entry into the map. And currently only get(key: Key, () -> Value): Value is protected by an explicit synchronized block to prevent calling the potentially expensive loader function multiple times concurrently. So I think those mutable fields in the CacheEntry should be the only things we need to protect from concurrent access and hence the volatile. Am I missing something here?

Reply:
Although using access to `ConcurrentHashMap` as a memory barrier should be enough, we do not write `Entry` back into the map after updating it so I don't believe we relay on it. If we wanted to drop `Volatile` here we would probably need to make these changes.

Whatever we end up with here (either volatile or using the map as a barrier), we should add some documentation here on why we're reusing entries (to be efficient) and document the concurrency policy if it's not obvious.

Reply:
Using `map.put(...)` as a barrier feels a bit subtle. Should we use an explicit lock to synchronize all updates to the `CacheEntry` objects?

Reply:
I've studied this a bit more and now I believe explicit lock or `volatile` shouldn't be necessary :)

Will remove `volatile` and add documentation about reuse.